src/hotspot/share/gc/g1/g1EvacuationFailureRegions.cpp

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2021, Huawei Technologies Co. Ltd. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "g1EvacuationFailureRegions.hpp"
+#include "gc/g1/g1CollectedHeap.hpp"
+#include "gc/g1/g1CollectedHeap.inline.hpp"
+#include "gc/g1/heapRegion.hpp"
+#include "runtime/atomic.hpp"
+
+
+G1EvacuationFailureRegions::G1EvacuationFailureRegions() :
+  _regions_failed_evacuation(mtGC) {
+}
+
+G1EvacuationFailureRegions::~G1EvacuationFailureRegions() {
+  FREE_C_HEAP_ARRAY(uint, _evac_failure_regions);
+}
+
+void G1EvacuationFailureRegions::initialize() {
+  Atomic::store(&_evac_failure_regions_cur_length, 0u);
+  _max_regions = G1CollectedHeap::heap()->max_reserved_regions();
+  _regions_failed_evacuation.resize(_max_regions);
+  _evac_failure_regions = NEW_C_HEAP_ARRAY(uint, _max_regions, mtGC);
+}
+
+void G1EvacuationFailureRegions::par_iterate(HeapRegionClosure* closure,
+                                             HeapRegionClaimer* _hrclaimer,
+                                             uint worker_id) {
+  assert_at_safepoint();
```

Comment:
As far as I can see this is a verbatim copy of `G1CollectionSet::iterate_part_from` (with some minor simplifications). Not really happy about that, need to think more about how/if we can avoid this.

Reply:
Agree. I think the reason is that we have mixed iteration itself with the logic in hrm/cset/evac failure regions.
One option to do might be to extract the iteration logic separately, should we do this in a separate issue?

Reply:
Actually, the collection set does not need the atomicity of addition to the array. So, probably it's best for this change to factor out the iteration over an array of heap region indexes into a `G1CollectedHeap`(probably the best place for now?) method which you pass the array information too - to be used in both places.

There is imho no point in delaying this into a separate CR, it's just one method and two uses afaict.

Reply:
I see your point. 
Did a very simple refactoring to move the iteration from G1CollectionSet  to G1CollectedHeap, and reuse this in both G1CollectionSet  and G1EvacFailureRegions.